File: dart/glue/lib/src/lib/bool/coalesce.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

final Ir coalesce = IrSpecial(coalesceImpl);

Eval<Ir> coalesceImpl(List<Ir> args) {
  return switch (args) {
    [final a, final b] => eval(a).bind((a_) {
      return isExist(a_) ? Eval.pure(a_) : eval(b);
    }),
    _ => throwError(wrongNumberOfArguments()),
  };
}

File: haskell/glue/src/Glue/Lib/Bool/Coalesce.hs
module Glue.Lib.Bool.Coalesce where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongNumberOfArguments)
import Glue.IR (IR (..), isExist)

coalesce :: IR Eval
coalesce = Special coalesceImpl

coalesceImpl :: [IR Eval] -> Eval (IR Eval)
coalesceImpl [a, b] = do
    a' <- eval a
    if isExist a'
        then pure a'
        else eval b
coalesceImpl _ = throwError wrongNumberOfArguments
